### PR TITLE
[Upsert] persist validDocsIndex snapshot for Pinot upsert optimization

### PR DESCRIPTION
**Pull request by deemoliu**
Part of a series of PRs for Pinot Upsert Optimizations.

Check [this doc](https://docs.google.com/document/d/124TSQ6ZRFRZd6uEAZpq98E68eBh0fmHYLvERzXtmKjk/edit#) out for Upsert TTL and Snapshot Design

Heap memory usage is one of the pain point for Pinot upsert. Upsert TTL support will help use cases with live session (each records only alive for a short period of time, callled TTL). With upsert TTL support, out-of-date primary keys can be removed from heap and numbers of active primary keys stored in heap will be reduced. 

TTL support will brings challenge when recovering validDocsIndexes after restart, because validDocIndexes requires to have full history of metadata data. After out-of-dated primarys get removed, validDocIndexes can not be recovered easily.

This PR address the challenging requirement by persisting and loading validDocSnapshots to disk to recover validDocIndex.

### CODE REVIEW

**Comment by deemoliu** [comment]
Will fix the checkstyle, test failures, and add unit tests

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java [hunk 1]
```diff
@@ -215,6 +216,13 @@ public List<StarTreeV2> getStarTrees() {
     return _starTreeIndexContainer != null ? _starTreeIndexContainer.getStarTrees() : null;
   }
 
+  @Nullable
+  public void setValidDocIds(ImmutableRoaringBitmap validDocIds) {
+    for (int docId: validDocIds.toArray()) {
+      _validDocIds.add(docId);
```
Comment (KKcorps) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java:222]: We can use `MutableRoaringBitmap.add` operation here.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java [hunk 2]
```diff
@@ -183,8 +183,8 @@ protected void doInit() {
       Preconditions.checkState(schema != null, "Failed to find schema for table: %s", _tableNameWithType);
 
       _primaryKeyColumns = schema.getPrimaryKeyColumns();
-      Preconditions.checkState(!CollectionUtils.isEmpty(_primaryKeyColumns),
-          "Primary key columns must be configured for dedup");
+      Preconditions
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:186]: Can you please re-apply the [Pinot Style](https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup#setup-ide)?
I don't know why you get different auto-formatting results, but these changes are not related to the PR, and make it quite hard to review. Let's also revert the unrelated auto-formatting changes in other classes

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java [hunk 3]
```diff
@@ -416,15 +415,18 @@ private void buildDedupMeta(ImmutableSegmentImpl immutableSegment) {
...
+    Integer partitionGroupId = SegmentUtils
+        .getRealtimeSegmentPartitionId(segmentName, _tableNameWithType, _helixManager, _primaryKeyColumns.get(0));
+    Preconditions.checkNotNull(partitionGroupId, String
+        .format("PartitionGroupId is not available for segment: '%s' (upsert-enabled table: %s)", segmentName,
             _tableNameWithType));
     PartitionUpsertMetadataManager partitionUpsertMetadataManager =
         _tableUpsertMetadataManager.getOrCreatePartitionManager(partitionGroupId);
+
     ThreadSafeMutableRoaringBitmap validDocIds = new ThreadSafeMutableRoaringBitmap();
+    if (immutableSegment.getValidDocIds() != null) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:427]: I don't think this is the correct way to handle it. We should get the persisted valid doc id, and iterate over it to get the `recordInfoIterator`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java [hunk 4]
```diff
@@ -198,6 +201,13 @@ public void init(SegmentGeneratorConfig config, SegmentCreationDataSource dataSo
     // Create a temporary directory used in segment creation
     _tempIndexDir = new File(indexDir, "tmp-" + UUID.randomUUID());
     LOGGER.debug("tempIndexDir:{}", _tempIndexDir);
+
+    _upsertSnapshotEnabled = _config.isUpsertSnapshotEnabled();
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:205]: We don't have the valid doc id info during segment creation. The snapshot has to be taken periodically during consumption. In the current implementation, no valid doc is added

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java (hunk 2 above)
Comment (deemoliu) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:186]: gotcha, thanks @Jackie-Jiang. let me re-import the codestyle and revert these.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java (hunk 3 above)
Comment (deemoliu) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:428]: Gotcha, thanks for review.
For immutableSegmentImpl, it already has `validDocsIds` variable for keeping in memory states, I was think of loading the validDocsSnapshots from disk to `validDocsIds` variable directly. Is there any concern/risk on this?
I can also add another variable call validDocsSnapshots that loaded from disk. Iterate over validDocsSnapshots to get the recordInfoInteractor and re-generate validDocIds in PartitionUpsertMetadataManager (addSegment).

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java (hunk 4 above)
Comment (deemoliu) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:205]: Thanks @Jackie-Jiang i planned to create another PR for TTL-related config and logics. Do you think I should add that in the PR?

**Comment by deemoliu** [comment]
@Jackie-Jiang pointed that there is a potential issue cause inconsistent query result.
#9095 #9101

when reloading the segments, the invalidDocsIndexes can be empty (drop and recreate) for a short period of time. If the persisting snapshot triggered during that time, we will persist empty snapshot and get wrong data. will need to rebase the change and then walk through corner cases when persist the snapshot.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java (hunk 4 above)
Comment (yupeng9) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:205]: Should we do this in `build` method? I think this shall be enabled with TTL, as it applies only to the segments outside TTL, otherwise the valid doc id will keep changing

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java [hunk 5]
```diff
@@ -348,7 +348,7 @@ protected Map<String, String> getStreamConfigMap() {
       streamConfigMap.put(KafkaStreamConfigProperties.constructStreamProperty(
           KafkaStreamConfigProperties.HighLevelConsumer.KAFKA_HLC_ZK_CONNECTION_STRING), getKafkaZKAddress());
       streamConfigMap.put(KafkaStreamConfigProperties.constructStreamProperty(
-              KafkaStreamConfigProperties.HighLevelConsumer.KAFKA_HLC_BOOTSTRAP_SERVER),
+          KafkaStreamConfigProperties.HighLevelConsumer.KAFKA_HLC_BOOTSTRAP_SERVER),
```
Comment (yupeng9) [inline pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:351]: curious, do you use the right formatter?

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java [hunk 6]
```diff
@@ -142,7 +142,7 @@ public class OfflineClusterIntegrationTest extends BaseClusterIntegrationTestSet
   private static final String MAX_NUM_MULTI_VALUES_MAP_KEY = "maxNumMultiValuesMap";
   // TODO: This might lead to flaky test, as this disk size is not deterministic
   //       as it depends on the iteration order of a HashSet.
-  private static final int DISK_SIZE_IN_BYTES = 20796000;
+  private static final int DISK_SIZE_IN_BYTES = 20796348;
```
Comment (yupeng9) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java:145]: why change on this?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java [hunk 7]
```diff
@@ -213,6 +214,13 @@ public List<StarTreeV2> getStarTrees() {
...
```
Comment (yupeng9) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java:220]: this is not set, but add?

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java (hunk 5 above)
Comment (deemoliu) [reply pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:351]: @yupeng9 I re-imported but the codestyle seems has slight difference between each intellij versions.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java (hunk 6 above)
Comment (deemoliu) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java:145]: thanks @yupeng9 for review. 
I added "segment.use.ttl" properties in the metadata.
When load from disk, I use this properties from `segmentMetadataPropertiesConfiguration` to initialize [validDocSnapshot](https://github.com/apache/pinot/pull/9062/commits/6bfab718afadfaa7139de227dd27edf7d0a44b9e)

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java (hunk 7 above)
Comment (deemoliu) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java]: Thanks @yupeng9 

Previously i tried to set the validDocIndex directly from snapshot. so i iterate validDocId from immutableRoaringBitmap to ThreadSafeRoaringBitmap. I synced with Jackie and learned this is not correct, so I fix it in `[this commit](https://github.com/apache/pinot/pull/9062/commits/56fb2a6b629473f5a64eb945597092854a39c3ba)`

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java (hunk 3 above)
Comment (deemoliu) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java]: fixed [here](https://github.com/apache/pinot/pull/9062/commits/56fb2a6b629473f5a64eb945597092854a39c3ba)

**Comment by deemoliu** [comment]
@yupeng9 @Jackie-Jiang Do you have any suggestion on when to persist snapshot? I looked into partitionUpsertMetaDataManager and summarized the [scenarios for persisting snapshot](https://docs.google.com/document/d/1guYCag3VoE2148oAj8v1k714qmuhrKI1Z1HNJt2BLqQ/edit#)

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 8]
```diff
@@ -55,12 +55,16 @@ public enum Strategy {
...
+  private boolean _useSnapshot;
+
   @JsonCreator
   public UpsertConfig(@JsonProperty(value = "mode", required = true) Mode mode,
       @JsonProperty("partialUpsertStrategies") @Nullable Map<String, Strategy> partialUpsertStrategies,
       @JsonProperty("defaultPartialUpsertStrategy") @Nullable Strategy defaultPartialUpsertStrategy,
       @JsonProperty("comparisonColumn") @Nullable String comparisonColumn,
-      @JsonProperty("hashFunction") @Nullable HashFunction hashFunction) {
+      @JsonProperty("hashFunction") @Nullable HashFunction hashFunction,
+      @JsonProperty("useSnapshot") @Nullable boolean useSnapshot) {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:67]: `enableSnapshot` might be better

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 9]
```diff
@@ -55,12 +55,16 @@ public enum Strategy {
   @JsonPropertyDescription("Column for upsert comparison, default to time column")
   private final String _comparisonColumn;
 
+  @JsonPropertyDescription("whether use snapshot for fast upsert metadata recovery")
+  private boolean _useSnapshot;
+
   @JsonCreator
   public UpsertConfig(@JsonProperty(value = "mode", required = true) Mode mode,
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:62]: We already got too many keys in this config. Suggest making a PR similar to #8743 to clean them up first.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java [hunk 10]
```diff
@@ -403,6 +408,10 @@ public Map<String, ColumnMetadata> getColumnMetadataMap() {
     return _columnMetadataMap;
   }
 
+  public boolean isUpsertSnapshotEnabled() {
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java:411]: We should not rely on segment metadata to determine whether snapshot is enabled. The upsert config might change anytime, and we want to use/drop snapshot accordingly.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java [hunk 11]
```diff
@@ -213,6 +215,16 @@ public List<StarTreeV2> getStarTrees() {
     return _starTreeIndexContainer != null ? _starTreeIndexContainer.getStarTrees() : null;
   }
 
+  @Nullable
+  public void setValidDocSnapshots(ImmutableRoaringBitmap validDocSnapshots) {
+    _validDocSnapshots = validDocSnapshots;
+  }
+
+  @Nullable
+  public ImmutableRoaringBitmap getValidDocSnapshots() {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java:224]: (minor) Suggest renaming it to `getValidDocIdsSnapshot()`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java [hunk 12]
```diff
@@ -213,6 +215,16 @@ public List<StarTreeV2> getStarTrees() {
     return _starTreeIndexContainer != null ? _starTreeIndexContainer.getStarTrees() : null;
   }
 
+  @Nullable
+  public void setValidDocSnapshots(ImmutableRoaringBitmap validDocSnapshots) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java:219]: I feel a easier way to manage the snapshot is to add 3 methods into this class:
- `@Nullable ImmutableRoaringBitmap loadValidDocIdsSnapshot()`
- `void persistValidDocIdsSnapshot()`
- `void deleteValidDocIdsSnapshot()`

One scenario we want to handle is that when snapshot is enabled then disabled, we should be able to delete the existing snapshots.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java (hunk 10 above)
Comment (deemoliu) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java:411]: Gotcha, thanks for the suggestion! let me address this.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 9 above)
Comment (deemoliu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:62]: thanks @Jackie-Jiang. Will do it and rebase. 

one quick question, will this affect backward compatibility?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java (hunk 12 above)
Comment (deemoliu) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java:219]: gotcha, thanks for pointing out, let me address this.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java (hunk 11 above)
Comment (deemoliu) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java:224]: gotcha, thanks

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 9 above)
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:62]: We may deprecate the current constructor and add an empty constructor to keep backward compatibility

Comment (deemoliu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:62]: updated in https://github.com/apache/pinot/pull/9150

**Review by Jackie-Jiang — APPROVED** [review]

**Comment by navina** [comment]
@yupeng9 / @Jackie-Jiang / @deemoliu can you please provide access to the design doc - [this doc](https://docs.google.com/document/d/124TSQ6ZRFRZd6uEAZpq98E68eBh0fmHYLvERzXtmKjk/edit#) ?
